runtime/lua/vim/diagnostic.lua

```diff
@@ -1007,15 +1009,57 @@ local function next_diagnostic(search_forward, opts)
   -- Adjust row to be 0-indexed
   position[1] = position[1] - 1
 
+  -- Clamp column
+  position[2] = math.max(0, position[2])
+
   local wrap = if_nil(opts.wrap, true)
 
-  local diagnostics = get_diagnostics(bufnr, opts, true)
+  local diagnostics = {} ---@type vim.Diagnostic[]
+  ---ns -> extmark_id -> Diagnostic[]
+  local extmarks_index = {} ---@type table<integer, table<integer, vim.api.keyset.get_extmark_item>>
+  for namespace in pairs(diagnostic_cache[bufnr]) do
+    local ns = M.get_namespace(namespace)
+
+    local extmarks = api.nvim_buf_get_extmarks(
+      bufnr,
+      ns.user_data.location_ns,
+      { 0, 0 },
+      { -1, -1 },
+      {}
+    )
+
+    extmarks_index[namespace] = {}
+
+    for _, extmark in ipairs(extmarks) do
+      extmarks_index[namespace][extmark[1]] = extmark
+    end
+
+    local filtered_diagnostics =
+      filter_by_severity(opts.severity, diagnostic_cache[bufnr][namespace])
+
+    for _, diagnostic in ipairs(filtered_diagnostics) do
+      for _, extmark in ipairs(extmarks) do
+        if extmark[1] == diagnostic.extmark_id then
+          table.insert(diagnostics, diagnostic)
+        end
+      end
```

Comment:
This same thing is done in `open_float` as well. Not sure if there should be a separate table mapping extmark ids to diagnostics, which would need to be cleaned up when unloading a buffer or just extracting it into a separate function would be enough.